ironfish/src/memPool/recentlyEvictedCache.ts

```diff
@@ -0,0 +1,218 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+import { Logger } from '../logger'
+import { MetricsMonitor } from '../metrics'
+import { TransactionHash } from '../primitives/transaction'
+import { PriorityQueue } from './priorityQueue'
+
+interface EvictionQueueEntry {
+  hash: TransactionHash
+  feeRate: bigint
+}
+
+interface ExpireAtQueueEntry {
+  hash: TransactionHash
+  expireAtSequence: number
+}
+
+/**
+ * A cache to track transactions that have recently been evicted from the mempool.
+ *
+ * This is primarily a mechanism to prevent transactions from being re-added to the mempool
+ * within a reasonable duration if for some reason they are evicted (i.e. the mempool is full
+ * and the transaction is underpriced).
+ *
+ * When full, transactions are evicted in order of `feeRate` descending to make room for the
+ * new transaction. This order is chosen because these transactions most likely to be successfully
+ * re-introduced into the mempool
+ */
+export class RecentlyEvictedCache {
+  private readonly logger: Logger
+
+  private readonly metrics: MetricsMonitor
+
+  /**
+   * The maximum size of the cache in number of transactions.
+   */
+  public readonly maxSize: number
+
+  /**
+   * A priority queue of transaction hashes ordered by `feeRate` descending.
+   * This is the eviction order for the cache when full.
+   */
+  private readonly evictionQueue: PriorityQueue<EvictionQueueEntry>
+
+  /**
+   * A priority queue of transaction hashes ordered by their expiration sequence.
+   */
+  // TODO(holahula): this can just be a normal queue because if y is inserted after x, y cannot expire before x
```

Comment:
Is this true? If it's referencing the expiration sequence on the transaction, that's a user-set value, and there's no guarantee transactions will have the same interval (in other words, some users might set an expiration sequence 30 blocks out, and some users might set an expiration sequence 10 blocks out)

Reply:
It references the current block + maximum time a txn can spend in the cache. Usage of `expiration` in the comment is a bit ambigous as to whether it's expiration from mempool / evicted cache so I'll change that

Reply:
Resolving, I think I misunderstood that this isn't the expiration sequence set on the transaction, it's actually a sequence generated by the RecentlyEvictedCache.